### PR TITLE
Add css-slam and gulp-html-minifier for even more minification

### PR DESCRIPTION
**Pull request by azakus**
optimizePipe now has options
TODO: thread flags through from CLI

### CODE REVIEW

**Comment by azakus** [comment]
TODO 2: add tests

**Comment by azakus** [comment]
@justinfagnani PTAL

**Comment by abdonrd** [comment]
@azakus what is `css-slam`?

File: src/build/optimize-pipe.ts [hunk 1]
```diff
@@ -10,9 +10,51 @@
 
 import * as stream from 'stream';
 
+import * as gulpif from 'gulp-if';
```
Comment (justinfagnani) [inline src/build/optimize-pipe.ts]: style: extra newline + sort

File: src/build/optimize-pipe.ts [hunk 2]
```diff
@@ -10,9 +10,51 @@
...
+const htmlmin = require('gulp-html-minifier');
+
+export interface OptimizePipeOptions {
+  /**
+   * Enable HTML minification
+   *
+   * Options passed to gulp-html-minifier
+   */
+  html?: {
+    collapseWhitespace?: boolean
```
Comment (justinfagnani) [inline src/build/optimize-pipe.ts]: semi-colons?

File: src/build/optimize-pipe.ts [hunk 3]
```diff
@@ -10,9 +10,51 @@
...
+  let outpipe = stream;
+  if (options) {
+    if (options.js && options.js.minify) {
+      outpipe = outpipe.pipe(new UglifyTransform());
+    }
+    if (options.css && options.css.stripWhitespace) {
+      outpipe = outpipe.pipe(cssSlam());
+    }
+    if (options.html) {
+      outpipe = outpipe.pipe(gulpif((f) => minimatch(f.path, '*.html', {matchBase: true}), htmlmin(options.html)))
```
Comment (justinfagnani) [inline src/build/optimize-pipe.ts]: > 80 cols

**Comment by justinfagnani** [comment]
LGTM

**Comment by azakus** [comment]
@abdonrd https://github.com/polymerlabs/css-slam. I need to give it a README. Custom CSS parser + stringifier to make sure Polymer's css is parsed correctly.
I would use clean-css, but the amount of options I want to disable was too high.

File: src/build/optimize-pipe.ts (hunk 1 above)
Comment (azakus) [inline src/build/optimize-pipe.ts]: Both gulp-if and minimatch are third_party dependencies, and sorted. Not sure why it needs a newline.

Comment (justinfagnani) [inline src/build/optimize-pipe.ts]: yeah, I was complaining about the newline :)

Comment (azakus) [inline src/build/optimize-pipe.ts]: Oh, 'stream' is an internal-to-node module. I thought we wanted to separate those.

Comment (justinfagnani) [inline src/build/optimize-pipe.ts]: ah, no. I meant exteral vs internal to project. basically separating './'  and '../' imports

**Comment by abdonrd** [comment]
@azakus thanks for the explication!

**Pull request merged**